lib/HLSL/HLMatrixLowerPass.cpp

```diff
@@ -2330,9 +2330,66 @@ void HLMatrixLowerPass::DeleteDeadInsts() {
   m_inDeadInstsSet.clear();
 }
 
+// Iterate GV users, and transparently iterate address space cast users as well
+class GVUserIter {
+private:
+  Value::user_iterator GVUser;
+  Value::user_iterator AddrSpaceCastUser;
+
+public:
+  User *U = nullptr;
+public:
+  GVUserIter() {}
+  GVUserIter(Value *V)
+  {
+    if (!V->user_empty())
+      GVUser = V->user_begin();
+    Next();
+  }
+  GVUserIter(const GVUserIter &other) {
+    GVUser = other.GVUser;
+    AddrSpaceCastUser = other.AddrSpaceCastUser;
+    U = other.U;
+  }
+
+  User *Next() {
+    U = nullptr;
+    if (!AddrSpaceCastUser.atEnd()) {
+      U = *(AddrSpaceCastUser++);
+      return U;
+    }
+    while (!GVUser.atEnd()) {
+      U = *(GVUser++);
+      if (ConstantExpr *CE = dyn_cast<ConstantExpr>(U)) {
+        if (CE->user_empty())
+          continue;
+        if (CE->getOpcode() == Instruction::AddrSpaceCast) {
+          AddrSpaceCastUser = CE->user_begin();
+          U = *(AddrSpaceCastUser++);
+          return U;
+        }
+      }
+      return U;
+    }
+    U = nullptr;
+    return U;
+  }
+
+  GVUserIter begin() { return *this; }
+  GVUserIter end() { return GVUserIter(); }
+  bool operator==(const GVUserIter &other) {
+    return (U == other.U && GVUser == other.GVUser && AddrSpaceCastUser == other.AddrSpaceCastUser);
+  }
+  bool operator!=(const GVUserIter &other) { return !(*this == (other)); }
+  GVUserIter &operator++() { Next(); return *this; }
+  GVUserIter operator++(int) { auto tmp = *this; ++*this; return tmp; }
+  User *operator*() { return U; }
+  User *operator->() { return U; }
+};
+
 static bool OnlyUsedByMatrixLdSt(Value *V) {
   bool onlyLdSt = true;
-  for (User *user : V->users()) {
+  for (User *user : GVUserIter(V)) {
```

Comment:
BTW I'm not kidding when I say that I'm rewriting this file. There's no way I can merge this directly, but I'll figure out how to map it to the new implementation.